sklearn/preprocessing/tests/test_polynomial.py

```diff
@@ -788,6 +795,258 @@ def test_polynomial_features_csr_X_dim_edges(deg, dim, interaction_only):
+        interaction_only=interaction_only, include_bias=include_bias, degree=degree
+    )
+
+    # Calculate the number of combinations a-priori, and if needed check for
+    # the correct ValueError and terminate the test early.
+    num_combinations = pf._num_combinations(
+        n_features=n_features,
+        min_degree=0,
+        max_degree=degree,
+        interaction_only=pf.interaction_only,
+        include_bias=pf.include_bias,
+    )
+    if num_combinations > np.iinfo(np.intp).max:
+        msg = (
+            r"The output that would result from the current configuration would have"
+            r" \d* features which is too large to be indexed"
+        )
+        with pytest.raises(ValueError, match=msg):
+            pf.fit(X)
+        return
+
+    # On older versions of scipy, a bug occurs when an intermediate matrix in
+    # `to_stack` in `hstack` fits within int32 however would require int64 when
+    # combined with all previous matrices in `to_stack`.
+    has_bug = False
+    max_int32 = np.iinfo(np.int32).max
+    cumulative_size = n_features + include_bias
+    for deg in range(2, degree + 1):
+        max_indptr = _calc_total_nnz(X.indptr, interaction_only, deg)
+        max_indices = _calc_expanded_nnz(n_features, interaction_only, deg) - 1
+        cumulative_size += max_indices + 1
+        needs_int64 = max(max_indices, max_indptr) > max_int32
+        has_bug |= not needs_int64 and cumulative_size > max_int32
+
+    if sp_version < parse_version("1.8.0") and has_bug:
+        msg = r"In scipy versions `<1.8.0`, the function `scipy.sparse.hstack`"
+        with pytest.raises(ValueError, match=msg):
+            X_trans = pf.fit_transform(X)
+        return
+
+    # When `n_features>=65535`, `scipy.sparse.hstack` may not use the right
+    # dtype for representing indices and indptr if `n_features` is still
+    # small enough so that each block matrix's indices and indptr arrays
+    # can be represented with `np.int32`. We test `n_features==65535`
+    # since it is guaranteed to run into this bug.
+    if (
+        sp_version < parse_version("1.9.2")
+        and n_features == 65535
+        and degree == 2
+        and not interaction_only
+    ):
+        msg = r"In scipy versions `<1.9.2`, the function `scipy.sparse.hstack`"
+        with pytest.raises(ValueError, match=msg):
+            X_trans = pf.fit_transform(X)
+        return
+    X_trans = pf.fit_transform(X)
+
+    expected_dtype = np.int64 if num_combinations > np.iinfo(np.int32).max else np.int32
+    higher_order_count = (degree - 1) * int(not interaction_only)
+    expected_nnz = 1 + higher_order_count + int(include_bias)
+    assert X_trans.dtype == X.dtype
+    assert X_trans.shape == (1, pf.n_output_features_)
+    assert X_trans.indptr.dtype == X_trans.indices.dtype == expected_dtype
+    assert X_trans.nnz == expected_nnz
+    for idx in range(higher_order_count):
+        assert X_trans[0, target_indices[idx]] == pytest.approx(1.0)
```

Comment:
I do not think this checks for all the non-zero indices. Is there a way to quickly check all that the indices of all the non-zeros elements are correct?